### PR TITLE
[WIP][1/2 Torch] Use pure torch/pyg/ray in examples

### PR DESCRIPTION
**Pull request by mortid0**
- [x] Forked repo is synced with upstream -> github shows no code delta outside of the desired.
    https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork
- [x] Tests are passing? https://github.com/microsoft/DeepGNN/blob/main/CONTRIBUTING.md#run-tests
- [x] Changelog and documentation updated.
- [x] PR is labeled using the label menu on the right side.

Previous Behavior
----------------
We use custom code to create models for DeepGNN package. However, it wasn't used as much as we expected, and it is easier to maintain pure torch/PyG/ray modules

New Behavior
----------------
* All examples are updated to use either PyG or pure torch modules.
* Added new dataset MOOC for temporal graphs.
* Added new model example(TGN) for temporal graphs.
* Added bazel target to create python wheel for deepgnn: `bazel build -c dbg --config=linux //src/python/deepgnn:deepgnn_ge_wheel` - will produce a python wheel with debug binaries
* Wheel target is used to run e2e examples. Utilize `pycross_wheel_library` to install wheel above to run containeraized examples. No need to manually do this in CI jobs and can test all examples locally,

### CODE REVIEW

File: examples/pytorch/BUILD [hunk 1]
```diff
@@ -1,12 +1,114 @@
 # Copyright (c) Microsoft Corporation.
 # Licensed under the MIT License.
+load("@rules_python//python:defs.bzl", "py_binary")
+load("@pip_deps//:requirements.bzl", "requirement")
+load("@jvolkman_rules_pycross//pycross:defs.bzl", "pycross_wheel_library")
```
Comment (coledie) [inline examples/pytorch/BUILD:5]: Why is the extra dependency necessary vs just using py_binary?

File: requirements.txt [hunk 2]
```diff
@@ -1,265 +1,62 @@
...
-wheel==0.38.1
-wrapt==1.14.1
-yacs==0.1.8
-yarl==1.8.1
-zict==2.2.0
-zipp==3.9.0
+sympy==1.12
+tenacity==8.2.2
+threadpoolctl==3.2.0
+--extra-index-url https://download.pytorch.org/whl/cpu
```
Comment (coledie) [inline requirements.txt:54]: This belongs not in file or at top of file maybe.

File: examples/pytorch/requirements.txt [hunk 3]
```diff
@@ -6,4 +6,4 @@ ray-on-aml==0.2.4
 ray[train]==2.4.0
 torch-geometric==2.0.4
 torch-scatter==2.0.9
-torch-sparse==0.6.13
+torch-sparse==0.6.16
```
Comment (coledie) [inline examples/pytorch/requirements.txt:9]: Make these versions the same as the ones in requirements.txt, otherwise we can maybe remove this file since we are using bazel for everything now.

File: src/python/deepgnn/graph_engine/data/mooc.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+import numpy as np
+
+from deepgnn.graph_engine.snark.meta import BINARY_DATA_VERSION
+from deepgnn.graph_engine.snark.converter.writers import BinaryWriter
+from deepgnn.graph_engine.data.data_util import download_file
+from deepgnn.graph_engine.snark.local import Client
+
+
+class MOOC(Client):
+    """Temporal MOOC dataset.
```
Comment (coledie) [inline src/python/deepgnn/graph_engine/data/mooc.py:22]: More description please.

File: examples/pytorch/gat_pyg.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    accuracy = torch.tensor(
+        accuracy_score(
+            y_true=label.cpu(), y_pred=scores.argmax(dim=1).detach().cpu().numpy()
+        )
+    ).item()
+    assert accuracy == 0.828
+
+
+if __name__ == "__main__":
+    train_func(
```
Comment (coledie) [inline examples/pytorch/gat_pyg.py:168]: why is train_func not in main?

File: examples/pytorch/hetgnn/graph.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+from typing import Tuple, Union, Any
+
+from torch.utils.data import IterableDataset
+
+from deepgnn.graph_engine import Graph, SamplingStrategy
+
+_node_base_index = 1000000
+
+
+class MockGraph(Graph):
```
Comment (coledie) [inline examples/pytorch/hetgnn/graph.py:14]: It looks like this is not used?

File: examples/pytorch/hetgnn/graph.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        raise "Not needed for HetGNN"
+
+    def random_walk(self, *args, **kwargs):
+        raise "Not needed for HetGNN"
+
+    def sample_edges(self, *args, **kwargs):
+        raise "Not needed for HetGNN"
+
+
+class MockHetGnnFileNodeLoader(IterableDataset):
```
Comment (coledie) [inline examples/pytorch/hetgnn/graph.py:165]: Same as above.

File: examples/pytorch/hetgnn/main.py [hunk 8]
```diff
@@ -1,90 +1,663 @@
...
-        init_optimizer_fn=create_optimizer,
-        init_args_fn=init_args,
+
+def evaluate_classification_on_het_gnn(config):  # noqa: F811
+    random.seed(0)
+    train_num, test_num = a_class_cluster_feature_setting(config)
+    macroF1, microF1 = evaluation.evaluate_node_classification(
+        train_num, test_num, config
+    )
+
```
Comment (coledie) [inline examples/pytorch/hetgnn/main.py:625]: This file is almost 1k lines, I think some stuff is not used can be removed.

File: examples/pytorch/sage.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        fanouts=[10, 10],
+    )
+    model.train()
+
+    optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
+    g = MemoryGraph(config["data_dir"], partitions=[0])
+    train_dataset = np.loadtxt(f"{config['data_dir']}/train.nodes", dtype=np.int64)
+    eval_batch = np.loadtxt(f"{config['data_dir']}/test.nodes", dtype=np.int64)
+
+    for epoch in range(config["num_epochs"]):
```
Comment (coledie) [inline examples/pytorch/sage.py:184]: Why don't we hardcode these arguments instead of putting them in train_func config?

File: examples/pytorch/sage.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+        losses = []
+        np.random.shuffle(train_dataset)
+        for batch in np.split(
+            train_dataset, train_dataset.size // config["batch_size"]
+        ):
+            train_tuple = model.query(g, batch)
+            loss, score, label = model(
+                {k: torch.from_numpy(train_tuple[k]) for k in train_tuple}
```
Comment (coledie) [inline examples/pytorch/sage.py:192]: Why doesn't model.query return this dict instead of the loop here.

File: src/python/deepgnn/graph_engine/snark/local.py [hunk 11]
```diff
@@ -290,7 +290,7 @@ def edge_features(
         return self.graph.edge_features(
             np.copy(edges[:, 0]),
             np.copy(edges[:, 1]),
-            np.copy(edges[:, 2]),
+            np.array(edges[:, 2], dtype=np.int32),
```
Comment (coledie) [inline src/python/deepgnn/graph_engine/snark/local.py:293]: This will create a pointer to the original edges array I believe, maybe would be more inline to do np.copy(edges[:, 2]).to(np.int32)

File: src/python/deepgnn/graph_engine/data/ppi.py [hunk 12]
```diff
@@ -52,7 +52,7 @@ def __init__(self, output_dir: Optional[str] = None, num_partitions: int = 1):
         if self.output_dir is None:
             self.output_dir = os.path.join(tempfile.gettempdir(), self.GRAPH_NAME)
         self._build_graph(self.output_dir)
-        super().__init__(path=self.output_dir, partitions=[0])
+        super().__init__(path=self.output_dir, partitions=list(range(num_partitions)))
```
Comment (coledie) [inline src/python/deepgnn/graph_engine/data/ppi.py:55]: I think a lot of small stuff like this belongs in a separate PR, there are quite a few things like this.

File: src/python/BUILD [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+
+py_library(
+    name = "deepgnn",
+    # srcs = [
+    #     "__init__.py",
+    #     "arg_types.py",
+    #     "log_consts.py",
+    #     "logging_utils.py",
+    #     "train_types.py",
+    # ],
```
Comment (coledie) [inline src/python/BUILD:14]: commented

File: examples/pytorch/tgn.py [hunk 14]
```diff
@@ -0,0 +1,236 @@
+import os
```
Comment (coledie) [inline examples/pytorch/tgn.py:1]: Copyright and description.

File: examples/pytorch/tgn.py [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+from deepgnn.graph_engine.data.mooc import MOOC
+from deepgnn import get_logger
+
+device = torch.device("cpu")
+train_batch_size = 200
+num_neighbors = 10
+torch.manual_seed(42)
+
+
+class GraphAttentionEmbedding(torch.nn.Module):
```
Comment (coledie) [inline examples/pytorch/tgn.py:23]: Model docstring

**Review by coledie — CHANGES_REQUESTED** [review]

File: examples/pytorch/gat_ray.py [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+            "get_graph": get_graph,
+            "data_dir": cora.data_dir(),
+            "num_epochs": 51,
+            "feature_idx": 0,
+            "feature_dim": 1433,
+            "label_idx": 1,
+            "label_dim": 1,
+            "num_classes": 7,
+        },
+        scaling_config=ScalingConfig(num_workers=1),
```
Comment (mortid0) [inline examples/pytorch/gat_ray.py:200]: num_workers=2

File: examples/pytorch/sage.py (hunk 9 above)
Comment (mortid0) [reply examples/pytorch/sage.py:184]: Names are more descriptive. Also easier to refactor if needed, e.g. some cusomt graphs might have 500dim features.

File: examples/pytorch/sage.py (hunk 10 above)
Comment (mortid0) [reply examples/pytorch/sage.py:192]: It is easier to work with a single type of data in a function. Query for numpy arrays, forward for tensors and glue code in the optimizer loop.

File: examples/pytorch/hetgnn/graph.py (hunk 6 above)
Comment (mortid0) [reply examples/pytorch/hetgnn/graph.py:14]: It is used, line main.py:659

File: examples/pytorch/hetgnn/graph.py (hunk 7 above)
Comment (mortid0) [reply examples/pytorch/hetgnn/graph.py:165]: main.py:587

File: examples/pytorch/hetgnn/main.py (hunk 8 above)
Comment (mortid0) [reply examples/pytorch/hetgnn/main.py:625]: I think all of it is used. The main goal of this example is to show how to work with a pure python mock data.

Comment (coledie) [reply examples/pytorch/hetgnn/main.py:625]: Can you refactor into multiple files then, I dont think huge files are very useful.

File: requirements.txt (hunk 2 above)
Comment (mortid0) [reply requirements.txt:54]: why? It is needed for torch package, that is jut below it.

File: examples/pytorch/gat_pyg.py (hunk 5 above)
Comment (mortid0) [reply examples/pytorch/gat_pyg.py:168]: why does it have to be main? Please provide a reference to a style guide.

File: examples/pytorch/BUILD (hunk 1 above)
Comment (mortid0) [reply examples/pytorch/BUILD:5]: Because `py_binary` requires dependencies returning `PyInfo` interface and `py_wheel` doesn't implement it.

File: examples/pytorch/hetgnn/main.py (hunk 8 above)
Comment (mortid0) [reply examples/pytorch/hetgnn/main.py:625]: This file has 625 lines, 1k is a brave round up.
I don't see any reference to file size in [PEP8](https://peps.python.org/pep-0008/)

File: examples/pytorch/requirements.txt (hunk 3 above)
Comment (mortid0) [reply examples/pytorch/requirements.txt:9]: Yeah, this file is not needed. Thanks!

File: requirements.txt (hunk 2 above)
Comment (coledie) [reply requirements.txt:69]: Then the extra_index_url at the top of the file used for torch geometric should be moved down next to it also.

**Comment by mortid0** [comment]
superceeded by https://github.com/microsoft/DeepGNN/pull/219
